### ISSUE TITLE
Executing stored procedure during the setup/cleanup

### ISSUE THREAD

**Issue by Seddryck** [Seddryck/NBi#327]
Currently, NBi supports to execute a SQL batch stored in an external file but creating an external file is perhaps a bit too much if you just want to run a stored procedure. NBi should support a way to define a stored procedure (or any T-SQL) that will be executed during the setup.

**Comment by Seddryck** [comment]
Sounds a good item to start to work on NBi and implement new features.

**_How to:_**

Before anything, clone the project and start from the branch named ```develop```. 

0. Select a name for this new task: probably something such as ```database-command``` or ```dml-command``` or ```dml-statement``` (is a stored procedure a statement?). A ```query``` is expecting to return a result-set ; for this task it won't be the case so probably better to have another name.

1. Add this new element in the ```base-type.xsd```: could/should inherits from ```query-enhanced-type``` or another ```query-...-type``` (don't use ```query-type``` it's deprecated), perhaps better to create your own level of inheritance to not let people do stupid stuff (get the query from a report sounds weird in this use-case).
1. Add a new class in ```NBi.Xml.Decoration.Command``` probably inheriting from ```QueryXml``` (or any other abstraction) and implementing ```IDecorationCommand```. Yes, I know the methods ```GetConnectionString()``` and ```GetCommand()``` are ugly but you should use them, I'll remove them in another refactoring.
1. Add two tests in ```NBi.Testing.Unit.Xml``` to serialize and deserialize the newly created task.
1. Create a new class in ```NBi.Core.Query``` to handle the execution of the stored procedure or other command. It must implement ```IDecorationImplementation``` ... add tests
1. Update the class ```DecorationFactory``` in NBi.Core to support your new task. You'll probably need to create a new interface in NBi.Core and set the newly created class in ```NBi.Xml.Decoration.Command``` to implement this interface.
1. Add an acceptance test to ```NBi.Testing.Acceptance``` with your new decoration task (I think that it's possible to write into a few tables of the database I'm referring in most of my tests).

And I think it's done :-)

If you've any doubt feel free to ask before starting anything that sounds complex.

**Comment by NeilRoper** [comment]
Hi @Seddryck,

You mentioned above that a stored proc can be run using a sql batch file but I am struggling to find this in the NBi documentation or in the issues log. 

Say I want to get around the SSIS package execution constraint that it needs to be run on the same server as SSIS is installed, is it possible to run a sql batch as below, returning an execution_id and finally returning the execution result and asserting success?

-- Create the execution object
DECLARE @execution_id BIGINT
EXEC [SSISDB].[catalog].[create_execution] 
	@package_name = N'Package1.dtsx'
	, @project_name = N'Project1'
	, @folder_name = N'Folder1'
	, @use32bitruntime = False
	, @reference_id = NULL
	, @execution_id = @execution_id OUTPUT

-- System parameters
EXEC [SSISDB].[catalog].[set_execution_parameter_value] 
	@execution_id
	, @object_type = 50						-- System parameter
	, @parameter_name = N'SYNCHRONIZED'
	, @parameter_value = 1

-- Execute the package
EXEC [SSISDB].[catalog].[start_execution] @execution_id

Sorry if this is in the wrong thread but I didn't want to raise a new issue as I am not sure it really is an issue. Any help appreciated.

**Comment by Seddryck** [comment]
I'm aware that several users use this trick to remotely execute SSIS packages. 

It has always been in my plans to implement this into NBi but a first step was to move out the SQL Server specifics items to the new extension `NBi.SqlServer`. I added the issue to natively support this in this [repository](https://github.com/Seddryck/NBi.SqlServer/issues/1).

It could be a bit tricky at the moment to use the SQL Server specific items because the release 1.21 include the extraction of these components from the main model. Rollback to v1.20 if you've issues but I plan to fix this during next weeks.

**Comment by NeilRoper** [comment]
Hi @Seddryck,

Thanks for the response. I was actually just looking for where stored procedures can be currently executed in NBi, thanks.

**Comment by NeilRoper** [comment]
Ok, have found what I need here http://www.nbi.io/docs/compare-equivalence-resultsets/
